File: golang/42_trap/trap.go
/*
 * @Date: 2023-07-23
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-07-23
 * @FilePath: /algorithm/golang/42_trap/trap.go
 */

// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func trap(height []int) (ans int) {
	left, right := 0, len(height)-1
	leftMax, rightMax := 0, 0
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	for left < right {
		leftMax = max(leftMax, height[left])
		rightMax = max(rightMax, height[right])
		if height[left] < height[right] {
			ans += leftMax - height[left]
			left++
		} else {
			ans += rightMax - height[right]
			right--
		}
	}
	return
}

func main() {
	tests := []struct {
		height []int
		ans    int
	}{
		{[]int{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6},
		{[]int{4, 2, 0, 3, 2, 5}, 9},
	}

	for _, item := range tests {
		assert.Equal(&testing.T{}, trap(item.height), item.ans)
	}
}

File: cpp/42_trap/trap.cpp
/*
 * @Date: 2023-07-23
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-07-23
 * @FilePath: /algorithm/cpp/42_trap/trap.cpp
 */

#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int trap(vector<int>& height) {
    int ans = 0;
    int left = 0, right = height.size() - 1;
    int leftMax = 0, rightMax = 0;
    while (left < right) {
      leftMax = max(leftMax, height[left]);
      rightMax = max(rightMax, height[right]);
      if (height[left] < height[right]) {
        ans += leftMax - height[left];
        ++left;
      } else {
        ans += rightMax - height[right];
        --right;
      }
    }
    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, int>> tests{
      {{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6},
      {{4, 2, 0, 3, 2, 5}, 9},
  };

  for (auto& [height, ans] : tests) {
    assert(Solution().trap(height) == ans);
  }
}
